lighthouse-core/lib/network-recorder.js

```diff
@@ -162,91 +159,123 @@ class NetworkRecorder extends EventEmitter {
   /**
    * Listener for the DevTools SDK NetworkManager's RequestStarted event, which includes both
    * web socket and normal request creation.
-   * @param {{data: LH.WebInspector.NetworkRequest}} request
+   * @param {NetworkRequest} request
    * @private
    */
   onRequestStarted(request) {
-    this._records.push(request.data);
+    this._records.push(request);
+    this._recordsById.set(request.requestId, request);
+
     this._emitNetworkStatus();
   }
 
   /**
    * Listener for the DevTools SDK NetworkManager's RequestFinished event, which includes
    * request finish, failure, and redirect, as well as the closing of web sockets.
-   * @param {{data: LH.WebInspector.NetworkRequest}} request
+   * @param {NetworkRequest} request
    * @private
    */
   onRequestFinished(request) {
-    this.emit('requestloaded', request.data);
+    this.emit('requestloaded', request);
     this._emitNetworkStatus();
   }
 
-  // The below methods proxy network data into the DevTools SDK network layer.
-  // There are a few differences between the debugging protocol naming and
-  // the parameter naming used in NetworkManager. These are noted below.
+  // The below methods proxy network data into the NetworkRequest object which mimics the
+  // DevTools SDK network layer.
 
   /**
    * @param {LH.Crdp.Network.RequestWillBeSentEvent} data
    */
   onRequestWillBeSent(data) {
-    // NOTE: data.timestamp -> time, data.type -> resourceType
-    this.networkManager._dispatcher.requestWillBeSent(data.requestId,
-        data.frameId, data.loaderId, data.documentURL, data.request,
-        data.timestamp, data.wallTime, data.initiator, data.redirectResponse,
-        data.type);
+    const originalRequest = this._findRealRequest(data.requestId);
+    // This is a simple new request, create the NetworkRequest object and finish.
+    if (!originalRequest) {
+      const request = new NetworkRequest();
+      request.onRequestWillBeSent(data);
+      this.onRequestStarted(request);
+      return;
+    }
+
+    // TODO(phulce): log these to sentry?
+    if (!data.redirectResponse) {
+      return;
+    }
+
+    // On redirect, another requestWillBeSent message is fired for the same requestId.
+    // Update the previous network request and create a new one for the redirect.
+    const modifiedData = {
+      ...data,
+      // Copy over the initiator as well to match DevTools behavior
+      // TODO(phulce): abandon this DT hack and update Lantern graph to handle it
+      initiator: originalRequest._initiator,
+      requestId: `${originalRequest.requestId}:redirect`,
+    };
+    const redirectRequest = new NetworkRequest();
```

Comment:
bikeshed: `redirectRequest` to me sounds like the original request (it returned with instructions to redirect). Maybe `redirectedRequest`?

Reply:
done